Integrate ENS Record Setting in UI

juliopavila opened the pull request:
## **Description**:

closes #253 

### Problem:
Currently, there's no direct way in the UI for users to set their ENS record. Users have to manually add their publication hash to their ENS record, as shown in the example: [ENS Domain Example](https://app.ens.domains/spiterman.eth?tab=records). While they can access it at `tabula.gg/#/yourCustomEns`, and the app will automatically detect the ENS, the process could be more streamlined.

### Solution:
This PR integrates the `multicall` and `setText` methods from the ENS contract directly into our UI. Users can now easily set their ENS record by adding a button within the publications section. This button will execute all the necessary logic and explain to the user about the underlying processes.

### Changes:
1. **App.tsx**: Imported `EnsProvider` from `./services/ens/context`.
2. **WalletBadge.tsx**: Made several modifications including importing `useEnsContext` and refactoring the logic.
3. **EnsModal.tsx**: New component to handle the ENS modal UI.
4. **SettingSection.tsx**: Added logic to link the publication to the user's ENS name.
5. **ens.context.tsx**: New context for ENS operations.
6. **useEns.ts**: New hook to handle ENS-related operations.

### Additional Context:
The goal of this PR is to enhance the user experience by simplifying the setting of ENS records and clarifying what happens in the background.

---

### **Reviewer Checklist**:
- [ ] Code is clean and follows the project's coding standards.
- [ ] Functionality has been verified and is working as intended.
- [ ] UI/UX is intuitive and matches the design specs.
- [ ] All new dependencies have been declared.
- [ ] PR is linked to the related issue.

### **Screenshots**

<img width="1688" alt="image" src="https://github.com/gnosis/tabula/assets/19823989/03d649e8-f340-48de-b24e-6cf298e4eb71">
<img width="1688" alt="image" src="https://github.com/gnosis/tabula/assets/19823989/1acc5cc5-65db-4aee-b017-462f99982611">
<img width="1688" alt="image" src="https://github.com/gnosis/tabula/assets/19823989/ac5b234c-b083-4a4e-b21a-8c46c4821382">
<img width="1701" alt="image" src="https://github.com/gnosis/tabula/assets/19823989/a17a102a-8176-4068-864e-16389915fd28">
![image](https://github.com/gnosis/tabula/assets/19823989/4a5136b5-dd9f-4275-bb60-a7b4d86f919d)

auryn-macmillan commented:
Checked out the deploy preview, but it's not showing the new elements to set ENS names.
![image](https://github.com/gnosis/tabula/assets/8453294/ecfd8e59-c0a3-4443-9772-53a58ce1e9c4)

manboy-eth reviewed (commented):
Nice work, just one note:
We will probably need some way of getting the user to change chain to mainnet, if they are not already on mainnet, since the ENS record is always set on mainnet.

manboy-eth commented on packages/app/src/services/ens/hooks/useENS.ts:
```diff
@@ -0,0 +1,200 @@
+        const web3Provider = new ethers.providers.Web3Provider(provider)
+        const contract = new ethers.Contract(ensRegistry, abiPublicResolver, web3Provider)
+        const signer = web3Provider.getSigner()
+        const data = contract.interface.encodeFunctionData("multicall", [[textRecord]])
+
+        if (!data) {
+          openNotification({
+            message: "Failed to encode data for multicall.",
+            variant: "error",
+          })
+          setLoading(false)
+          return
+        }
+
+        const tx = await signer.sendTransaction({
+          to: publicResolver,
+          data: data,
+        })
+        const receipt: TransactionReceipt = await tx.wait()
+
+        openNotification({
+          message: "Transaction completed successfully!",
+          variant: "success",
+          detailsLink: `${URL}tx/${receipt.transactionHash}`,
+        })
+        setTransactionCompleted(true)
+      } catch (e) {
+        console.log("error", e)
+        setLoading(false)
+
+        openNotification({
+          message: "ENS is not supported on this network or an error occurred.",
+          variant: "error",
+        })
+      } finally {
+        setLoading(false)
+      }
+    },
+    [getPublicResolverAddress, openNotification],
+  )
```
Why do we do a 'multicall' when we are only doing one call?

cedricwaxwing reviewed (approved):
LGTM, nice work Julio!

juliopavila commented:
Thanks, for your feedback @manboy-eth, I already applied the following changes:

- Changed default to mainnet subgraph, the only "real" ENS registry, as per your suggestion.
- Replaced 'multicall' with 'setText' for a single call operation, cleaned up the code, and tested functionality.

Please review

manboy-eth reviewed (commented):
Nice, much better :)

But did you do anything to handle this?
> We will probably need some way of getting the user to change chain to mainnet, if they are not already on mainnet, since the ENS record is always set on mainnet.

When the user for instance has a publication on Gnosis Chain, the user must still set the ENS record on mainnet.

manboy-eth commented on packages/app/src/components/commons/NetworkModal.tsx:
```diff
@@ -0,0 +1,111 @@
+  padding: 24,
+})
+
+const NetworkModal: React.FC<any> = ({ open, handleClose }) => {
```
Missing props type.

manboy-eth commented on packages/app/src/components/commons/NetworkModal.tsx:
```diff
@@ -0,0 +1,111 @@
+      try {
+        setLoading(true)
+        await library.send("wallet_switchEthereumChain", [{ chainId: "0x1" }])
+        handleClose({}, "escapeKeyDown")
```
Used as MUIs `onClose` ModalProp, but looks like it is actually a `() => void` (from component usage in `packages/app/src/components/views/publication/components/SettingSection.tsx`).

manboy-eth reviewed (changes requested):
Excellent, now we just need some minor changes and I think this is ready to merge.
